File: cobble-binding/cobble-cpp/cpp/structured/multi_get.cpp
#include <cobble/structured/db.hpp>
#include <cobble/structured/multi_get.hpp>
#include <cobble/structured/single_db.hpp>

#include <cstddef>
#include <cstdint>
#include <limits>
#include <type_traits>
#include <vector>

#include "../detail/error.hpp"
#include "detail/convert.hpp"
#include "detail/impl.hpp"

namespace cobble::structured {
namespace {

constexpr std::size_t AlignUp(std::size_t value, std::size_t alignment) {
  return (value + alignment - 1) / alignment * alignment;
}

struct KeyDescriptor {
  std::uint16_t bucket;
  std::uint16_t reserved;
  const Byte *data;
  std::size_t length;
};

constexpr std::size_t kPointerOffset =
    AlignUp(sizeof(std::uint16_t) * 2, alignof(const Byte *));
constexpr std::size_t kLengthOffset =
    AlignUp(kPointerOffset + sizeof(const Byte *), alignof(std::size_t));
constexpr std::size_t kDescriptorAlignment = alignof(const Byte *) >
                                                     alignof(std::size_t)
                                                 ? alignof(const Byte *)
                                                 : alignof(std::size_t);
constexpr std::size_t kDescriptorSize =
    AlignUp(kLengthOffset + sizeof(std::size_t), kDescriptorAlignment);

static_assert(std::is_standard_layout_v<KeyDescriptor>);
static_assert(offsetof(KeyDescriptor, bucket) == 0);
static_assert(offsetof(KeyDescriptor, reserved) == sizeof(std::uint16_t));
static_assert(offsetof(KeyDescriptor, data) == kPointerOffset);
static_assert(offsetof(KeyDescriptor, length) == kLengthOffset);
static_assert(alignof(KeyDescriptor) == kDescriptorAlignment);
static_assert(sizeof(KeyDescriptor) == kDescriptorSize);

std::uint64_t Count(std::size_t value, const char *name) {
  if (value > std::numeric_limits<std::uint64_t>::max()) {
    throw Error(ErrorCode::kInput, std::string(name) + " exceeds uint64_t");
  }
  return static_cast<std::uint64_t>(value);
}

std::vector<KeyDescriptor> Descriptors(std::span<const MultiGetKey> keys) {
  std::vector<KeyDescriptor> descriptors;
  descriptors.reserve(keys.size());
  for (const auto &key : keys) {
    descriptors.push_back({key.bucket, 0, key.key.data(), key.key.size()});
  }
  return descriptors;
}

std::size_t Address(const std::vector<KeyDescriptor> &values) noexcept {
  return values.empty() ? 0 : reinterpret_cast<std::size_t>(values.data());
}

template <typename T>
void Require(const std::shared_ptr<T> &impl, const char *message) {
  if (!impl) {
    throw Error(ErrorCode::kInvalidState, message);
  }
}

} // namespace

OwnedMultiGetResult::OwnedMultiGetResult(std::unique_ptr<Impl> impl) noexcept
    : impl_(std::move(impl)) {}
OwnedMultiGetResult::OwnedMultiGetResult(OwnedMultiGetResult &&) noexcept =
    default;
OwnedMultiGetResult &
OwnedMultiGetResult::operator=(OwnedMultiGetResult &&) noexcept = default;
OwnedMultiGetResult::~OwnedMultiGetResult() = default;

std::size_t OwnedMultiGetResult::RowCount() const noexcept {
  return impl_ ? structured_ffi::native_structured_multi_get_row_count(
                     *impl_->native)
               : 0;
}

bool OwnedMultiGetResult::Found(std::size_t row) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  return structured_ffi::native_structured_multi_get_found(*impl_->native, row);
}

std::size_t OwnedMultiGetResult::ColumnCount(std::size_t row) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  return detail::Translate([&] {
    return structured_ffi::native_structured_multi_get_column_count(
        *impl_->native, row);
  });
}

bool OwnedMultiGetResult::HasColumn(std::size_t row, std::size_t column) const {
  if (!impl_)
    return false;
  return structured_ffi::native_structured_multi_get_has_column(*impl_->native,
                                                                row, column);
}

ColumnKind OwnedMultiGetResult::Kind(std::size_t row,
                                     std::size_t column) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  return static_cast<ColumnKind>(detail::Translate([&] {
    return structured_ffi::native_structured_multi_get_kind(*impl_->native, row,
                                                            column);
  }));
}

BytesView OwnedMultiGetResult::Bytes(std::size_t row,
                                     std::size_t column) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  return detail::ToView(detail::Translate([&] {
    return structured_ffi::native_structured_multi_get_bytes(*impl_->native,
                                                             row, column);
  }));
}

std::size_t OwnedMultiGetResult::ListSize(std::size_t row,
                                          std::size_t column) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  return detail::Translate([&] {
    return structured_ffi::native_structured_multi_get_list_size(*impl_->native,
                                                                 row, column);
  });
}

BytesView OwnedMultiGetResult::ListElement(std::size_t row, std::size_t column,
                                           std::size_t element) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState,
                "OwnedMultiGetResult has been moved from");
  return detail::ToView(detail::Translate([&] {
    return structured_ffi::native_structured_multi_get_list_element(
        *impl_->native, row, column, element);
  }));
}

BufferResult Db::GetInto(BucketId bucket, BytesView key,
                         MutableBytesView output,
                         const ReadOptions &options) const {
  Require(impl_, "structured Db has been moved from");
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ReadOptions has been moved from");
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_db_get_into(
        *impl_->native, bucket, detail::RustBytes(key), *options.impl_->native,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

BufferResult Db::GetInto(BucketId bucket, BytesView key,
                         MutableBytesView output) const {
  const ReadOptions options;
  return GetInto(bucket, key, output, options);
}

OwnedMultiGetResult Db::MultiGet(std::span<const MultiGetKey> keys,
                                 const ReadOptions &options) const {
  Require(impl_, "structured Db has been moved from");
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ReadOptions has been moved from");
  const auto descriptors = Descriptors(keys);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_db_multi_get(
        *impl_->native, Address(descriptors), Count(keys.size(), "key count"),
        *options.impl_->native);
  });
  return OwnedMultiGetResult(
      std::make_unique<OwnedMultiGetResult::Impl>(std::move(native)));
}

OwnedMultiGetResult Db::MultiGet(std::span<const MultiGetKey> keys) const {
  const ReadOptions options;
  return MultiGet(keys, options);
}

BufferResult Db::MultiGetInto(std::span<const MultiGetKey> keys,
                              MutableBytesView output,
                              const ReadOptions &options) const {
  Require(impl_, "structured Db has been moved from");
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ReadOptions has been moved from");
  const auto descriptors = Descriptors(keys);
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_db_multi_get_into(
        *impl_->native, Address(descriptors), Count(keys.size(), "key count"),
        *options.impl_->native,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

BufferResult Db::MultiGetInto(std::span<const MultiGetKey> keys,
                              MutableBytesView output) const {
  const ReadOptions options;
  return MultiGetInto(keys, output, options);
}

BufferResult SingleDb::GetInto(BucketId bucket, BytesView key,
                               MutableBytesView output,
                               const ReadOptions &options) const {
  Require(impl_, "structured SingleDb has been moved from");
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ReadOptions has been moved from");
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_single_db_get_into(
        *impl_->native, bucket, detail::RustBytes(key), *options.impl_->native,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

BufferResult SingleDb::GetInto(BucketId bucket, BytesView key,
                               MutableBytesView output) const {
  const ReadOptions options;
  return GetInto(bucket, key, output, options);
}

OwnedMultiGetResult SingleDb::MultiGet(std::span<const MultiGetKey> keys,
                                       const ReadOptions &options) const {
  Require(impl_, "structured SingleDb has been moved from");
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ReadOptions has been moved from");
  const auto descriptors = Descriptors(keys);
  auto native = detail::Translate([&] {
    return structured_ffi::native_structured_single_db_multi_get(
        *impl_->native, Address(descriptors), Count(keys.size(), "key count"),
        *options.impl_->native);
  });
  return OwnedMultiGetResult(
      std::make_unique<OwnedMultiGetResult::Impl>(std::move(native)));
}

OwnedMultiGetResult
SingleDb::MultiGet(std::span<const MultiGetKey> keys) const {
  const ReadOptions options;
  return MultiGet(keys, options);
}

BufferResult SingleDb::MultiGetInto(std::span<const MultiGetKey> keys,
                                    MutableBytesView output,
                                    const ReadOptions &options) const {
  Require(impl_, "structured SingleDb has been moved from");
  if (!options.impl_)
    throw Error(ErrorCode::kInvalidState, "ReadOptions has been moved from");
  const auto descriptors = Descriptors(keys);
  return detail::ToBufferResult(detail::Translate([&] {
    return structured_ffi::native_structured_single_db_multi_get_into(
        *impl_->native, Address(descriptors), Count(keys.size(), "key count"),
        *options.impl_->native,
        rust::Slice<Byte>(output.data(), output.size()));
  }));
}

BufferResult SingleDb::MultiGetInto(std::span<const MultiGetKey> keys,
                                    MutableBytesView output) const {
  const ReadOptions options;
  return MultiGetInto(keys, output, options);
}

} // namespace cobble::structured

File: cobble-binding/cobble-cpp/src/structured/multi_get.rs
use std::{mem, slice};

use cobble_binding::structured::StructuredColumnValue;

use crate::structured_bridge::ffi;

use super::conversion::{format_error, input_error};
use super::encoding::{
    CsrbColumns, CsrbRow, STATUS_BUFFER_TOO_SMALL, STATUS_OK, encode_into, encoded_len,
};
use super::{
    BridgeResult, NativeStructuredDb, NativeStructuredReadOptions, NativeStructuredSingleDb,
};

#[repr(C)]
struct KeyDescriptor {
    bucket: u16,
    reserved: u16,
    data: *const u8,
    length: usize,
}

pub(crate) struct NativeStructuredMultiGetResult {
    rows: Vec<Option<Vec<Option<StructuredColumnValue>>>>,
}

/// Borrows all key payloads for one synchronous structured multi-get crossing.
///
/// # Safety
///
/// The descriptor address and every non-empty key pointer must refer to readable caller memory
/// that remains alive until this function returns.
pub(crate) unsafe fn borrowed_keys<'a>(
    descriptor_address: usize,
    count: u64,
) -> BridgeResult<Vec<(u16, &'a [u8])>> {
    let count = usize::try_from(count).map_err(|_| input_error("key count exceeds usize"))?;
    if count == 0 {
        return Ok(Vec::new());
    }
    if descriptor_address == 0 {
        return Err(input_error("key descriptors must not be null"));
    }
    if !descriptor_address.is_multiple_of(mem::align_of::<KeyDescriptor>()) {
        return Err(input_error("key descriptors are not aligned"));
    }
    let bytes = count
        .checked_mul(mem::size_of::<KeyDescriptor>())
        .ok_or_else(|| input_error("key descriptor size overflows"))?;
    if bytes > isize::MAX as usize {
        return Err(input_error("key descriptor array exceeds isize::MAX"));
    }
    descriptor_address
        .checked_add(bytes)
        .ok_or_else(|| input_error("key descriptor address overflows"))?;
    // SAFETY: caller contract plus the array checks above.
    let descriptors =
        unsafe { slice::from_raw_parts(descriptor_address as *const KeyDescriptor, count) };
    descriptors
        .iter()
        .map(|descriptor| {
            if descriptor.reserved != 0 {
                return Err(input_error("key descriptor reserved field must be zero"));
            }
            if descriptor.length == 0 {
                return Ok((descriptor.bucket, &[][..]));
            }
            if descriptor.data.is_null() {
                return Err(input_error("non-empty key data must not be null"));
            }
            if descriptor.length > isize::MAX as usize {
                return Err(input_error("key length exceeds isize::MAX"));
            }
            (descriptor.data as usize)
                .checked_add(descriptor.length)
                .ok_or_else(|| input_error("key address overflows"))?;
            // SAFETY: caller keeps each readable key span alive for this synchronous call.
            Ok((descriptor.bucket, unsafe {
                slice::from_raw_parts(descriptor.data, descriptor.length)
            }))
        })
        .collect()
}

pub(crate) fn native_structured_db_multi_get(
    db: &NativeStructuredDb,
    descriptor_address: usize,
    count: u64,
    options: &NativeStructuredReadOptions,
) -> BridgeResult<Box<NativeStructuredMultiGetResult>> {
    // SAFETY: private C++ wrapper owns and validates the exact descriptor layout.
    let keys = unsafe { borrowed_keys(descriptor_address, count)? };
    db.db
        .multi_get_with_options(&keys, &options.options)
        .map(|rows| Box::new(NativeStructuredMultiGetResult { rows }))
        .map_err(format_error)
}

pub(crate) fn native_structured_single_db_multi_get(
    db: &NativeStructuredSingleDb,
    descriptor_address: usize,
    count: u64,
    options: &NativeStructuredReadOptions,
) -> BridgeResult<Box<NativeStructuredMultiGetResult>> {
    // SAFETY: private C++ wrapper owns and validates the exact descriptor layout.
    let keys = unsafe { borrowed_keys(descriptor_address, count)? };
    db.db
        .multi_get_with_options(&keys, &options.options)
        .map(|rows| Box::new(NativeStructuredMultiGetResult { rows }))
        .map_err(format_error)
}

pub(crate) fn native_structured_db_multi_get_into(
    db: &NativeStructuredDb,
    descriptor_address: usize,
    count: u64,
    options: &NativeStructuredReadOptions,
    output: &mut [u8],
) -> BridgeResult<ffi::NativeBufferResult> {
    // SAFETY: private C++ wrapper owns and validates the exact descriptor layout.
    let keys = unsafe { borrowed_keys(descriptor_address, count)? };
    let rows = db
        .db
        .multi_get_with_options(&keys, &options.options)
        .map_err(format_error)?;
    encode_multi_get(&keys, &rows, output)
}

pub(crate) fn native_structured_single_db_multi_get_into(
    db: &NativeStructuredSingleDb,
    descriptor_address: usize,
    count: u64,
    options: &NativeStructuredReadOptions,
    output: &mut [u8],
) -> BridgeResult<ffi::NativeBufferResult> {
    // SAFETY: private C++ wrapper owns and validates the exact descriptor layout.
    let keys = unsafe { borrowed_keys(descriptor_address, count)? };
    let rows = db
        .db
        .multi_get_with_options(&keys, &options.options)
        .map_err(format_error)?;
    encode_multi_get(&keys, &rows, output)
}

fn encode_multi_get(
    keys: &[(u16, &[u8])],
    rows: &[Option<Vec<Option<StructuredColumnValue>>>],
    output: &mut [u8],
) -> BridgeResult<ffi::NativeBufferResult> {
    if keys.len() != rows.len() {
        return Err(input_error("structured multi-get result count mismatch"));
    }
    let encoded_rows = keys
        .iter()
        .zip(rows)
        .map(|((bucket, key), columns)| CsrbRow {
            bucket: *bucket,
            key,
            columns: columns
                .as_deref()
                .map(CsrbColumns::Structured)
                .unwrap_or(CsrbColumns::Missing),
        })
        .collect::<Vec<_>>();
    let required = encoded_len(&encoded_rows)?;
    if output.len() < required {
        return Ok(buffer_result(
            STATUS_BUFFER_TOO_SMALL,
            0,
            required,
            rows.len(),
        ));
    }
    let written = encode_into(&encoded_rows, false, false, output)?;
    Ok(buffer_result(STATUS_OK, written, written, rows.len()))
}

fn row(
    result: &NativeStructuredMultiGetResult,
    index: usize,
) -> BridgeResult<&Option<Vec<Option<StructuredColumnValue>>>> {
    result
        .rows
        .get(index)
        .ok_or_else(|| input_error("multi-get row index is out of bounds"))
}

fn column(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
    column_index: usize,
) -> BridgeResult<&StructuredColumnValue> {
    row(result, row_index)?
        .as_ref()
        .and_then(|columns| columns.get(column_index))
        .and_then(Option::as_ref)
        .ok_or_else(|| input_error("multi-get column is absent"))
}

pub(crate) fn native_structured_multi_get_row_count(
    result: &NativeStructuredMultiGetResult,
) -> usize {
    result.rows.len()
}

pub(crate) fn native_structured_multi_get_found(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
) -> bool {
    row(result, row_index).ok().is_some_and(Option::is_some)
}

pub(crate) fn native_structured_multi_get_column_count(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
) -> BridgeResult<usize> {
    Ok(row(result, row_index)?.as_ref().map_or(0, Vec::len))
}

pub(crate) fn native_structured_multi_get_has_column(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
    column_index: usize,
) -> bool {
    column(result, row_index, column_index).is_ok()
}

pub(crate) fn native_structured_multi_get_kind(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
    column_index: usize,
) -> BridgeResult<u8> {
    Ok(match column(result, row_index, column_index)? {
        StructuredColumnValue::Bytes(_) => 0,
        StructuredColumnValue::List(_) => 1,
    })
}

pub(crate) fn native_structured_multi_get_bytes(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
    column_index: usize,
) -> BridgeResult<&[u8]> {
    match column(result, row_index, column_index)? {
        StructuredColumnValue::Bytes(value) => Ok(value),
        StructuredColumnValue::List(_) => Err(input_error("multi-get column is not BYTES")),
    }
}

pub(crate) fn native_structured_multi_get_list_size(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
    column_index: usize,
) -> BridgeResult<usize> {
    match column(result, row_index, column_index)? {
        StructuredColumnValue::List(elements) => Ok(elements.len()),
        StructuredColumnValue::Bytes(_) => Err(input_error("multi-get column is not LIST")),
    }
}

pub(crate) fn native_structured_multi_get_list_element(
    result: &NativeStructuredMultiGetResult,
    row_index: usize,
    column_index: usize,
    element_index: usize,
) -> BridgeResult<&[u8]> {
    match column(result, row_index, column_index)? {
        StructuredColumnValue::List(elements) => elements
            .get(element_index)
            .map(AsRef::as_ref)
            .ok_or_else(|| input_error("LIST element index is out of bounds")),
        StructuredColumnValue::Bytes(_) => Err(input_error("multi-get column is not LIST")),
    }
}

pub(crate) fn buffer_result(
    status: u8,
    bytes_written: usize,
    bytes_required: usize,
    row_count: usize,
) -> ffi::NativeBufferResult {
    ffi::NativeBufferResult {
        status,
        bytes_written: bytes_written as u64,
        bytes_required: bytes_required as u64,
        row_count: row_count as u64,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn descriptor_validation_rejects_null_misaligned_and_reserved() {
        // SAFETY: all cases are rejected before an invalid pointer is dereferenced.
        unsafe {
            assert!(
                borrowed_keys(0, 1)
                    .unwrap_err()
                    .contains("must not be null")
            );
            assert!(borrowed_keys(1, 1).unwrap_err().contains("not aligned"));
        }
        let descriptor = KeyDescriptor {
            bucket: 0,
            reserved: 1,
            data: std::ptr::null(),
            length: 0,
        };
        // SAFETY: descriptor points to a live, correctly aligned local value.
        let error = unsafe { borrowed_keys((&descriptor as *const KeyDescriptor) as usize, 1) }
            .unwrap_err();
        assert!(error.contains("reserved"));
    }

    #[test]
    fn descriptor_validation_accepts_empty_and_binary_keys() {
        let bytes = [0, 0xff, 1];
        let descriptors = [
            KeyDescriptor {
                bucket: 2,
                reserved: 0,
                data: std::ptr::null(),
                length: 0,
            },
            KeyDescriptor {
                bucket: 3,
                reserved: 0,
                data: bytes.as_ptr(),
                length: bytes.len(),
            },
        ];
        // SAFETY: descriptors and key payload remain live for the call.
        let keys = unsafe { borrowed_keys(descriptors.as_ptr() as usize, 2) }.unwrap();
        assert_eq!(keys, vec![(2, &[][..]), (3, &bytes[..])]);
    }
}
